homeassistant/components/tuya/__init__.py

```diff
@@ -0,0 +1,290 @@
+    hass.data[DOMAIN][TUYA_HOME_MANAGER] = home_manager
+
+    class DeviceListener(TuyaDeviceListener):
+        """Device Update Listener."""
+
+        def update_device(self, device: TuyaDevice):
+            for ha_device in hass.data[DOMAIN][TUYA_HA_DEVICES]:
+                if ha_device.tuya_device.id == device.id:
+                    _LOGGER.debug(f"_update-->{self};->>{ha_device.tuya_device.status}")
+                    ha_device.schedule_update_ha_state()
+
+        def add_device(self, device: TuyaDevice):
+
+            device_add = False
+
+            _LOGGER.info(
+                f"""add device category->{device.category}; keys->,
+                {hass.data[DOMAIN][TUYA_HA_TUYA_MAP].keys()}"""
+            )
+            if device.category in itertools.chain(
+                *hass.data[DOMAIN][TUYA_HA_TUYA_MAP].values()
+            ):
+                ha_tuya_map = hass.data[DOMAIN][TUYA_HA_TUYA_MAP]
+
+                remove_hass_device(hass, device.id)
+
+                for key, tuya_list in ha_tuya_map.items():
+                    if device.category in tuya_list:
+                        device_add = True
+                        async_dispatcher_send(
+                            hass, TUYA_DISCOVERY_NEW.format(key), [device.id]
+                        )
+
+            if device_add:
+                device_manager = hass.data[DOMAIN][TUYA_DEVICE_MANAGER]
+                device_manager.mq.stop()
+                tuya_mq = TuyaOpenMQ(device_manager.api)
+                tuya_mq.start()
+
+                device_manager.mq = tuya_mq
+                tuya_mq.add_message_listener(device_manager._on_message)
+
+        def remove_device(self, device_id: str):
+            _LOGGER.info(f"tuya remove device:{device_id}")
+            remove_hass_device(hass, device_id)
+
+    __listener = DeviceListener()
+    hass.data[DOMAIN][TUYA_MQTT_LISTENER] = __listener
+    device_manager.add_device_listener(__listener)
+    hass.data[DOMAIN][TUYA_DEVICE_MANAGER] = device_manager
+
+    # Clean up device entities
+    await cleanup_device_registry(hass)
+
+    _LOGGER.info(f"init support type->{TUYA_SUPPORT_HA_TYPE}")
+
+    for platform in TUYA_SUPPORT_HA_TYPE:
+        _LOGGER.info(f"tuya async platform-->{platform}")
+        hass.async_create_task(
+            hass.config_entries.async_forward_entry_setup(entry, platform)
+        )
+        hass.data[DOMAIN][TUYA_SETUP_PLATFORM].add(platform)
+
+    return True
+
+
+async def cleanup_device_registry(hass: HomeAssistant):
+    """Remove deleted device registry entry if there are no remaining entities."""
+
+    device_registry = hass.helpers.device_registry.async_get(hass)
+    device_manager = hass.data[DOMAIN][TUYA_DEVICE_MANAGER]
+
+    for dev_id, device_entity in list(device_registry.devices.items()):
+        for item in device_entity.identifiers:
+            if DOMAIN == item[0] and item[1] not in device_manager.device_map.keys():
```

Comment:
`keys()` is implicit.
```suggestion
            if DOMAIN == item[0] and item[1] not in device_manager.device_map:
```

Reply:
fixed.